server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -172,7 +176,57 @@ boolean isJoinPending() {
         return pendingOutgoingJoins.iterator().hasNext();
     }
 
-    void sendJoinRequest(DiscoveryNode destination, Optional<Join> optionalJoin) {
+    private static class FailedJoinAttempt {
+        private final DiscoveryNode destination;
+        private final JoinRequest joinRequest;
+        private final TransportException exception;
+        private final long timestamp;
+
+        FailedJoinAttempt(DiscoveryNode destination, JoinRequest joinRequest, TransportException exception) {
+            this.destination = destination;
+            this.joinRequest = joinRequest;
+            this.exception = exception;
+            this.timestamp = System.nanoTime();
+        }
+
+        void maybeLogNow() {
+            if (isSuspiciousTransportException(exception)) {
+                logger.info(() -> new ParameterizedMessage("failed to join {} with {}", destination, joinRequest), exception);
+            } else {
+                logger.debug(() -> new ParameterizedMessage("failed to join {} with {}", destination, joinRequest), exception);
+            }
+        }
+
+        boolean isSuspiciousTransportException(TransportException e) {
```

Comment:
Can this be `static`? Also, perhaps return the `Level` and then you can just call `logger.log(isSuspiciousTransportException(exception),...`. Also, this is worthy of a test.

Reply:
Done as a part of this commit https://github.com/elastic/elasticsearch/pull/39756/commits/737e7f8fb00b4f69093780e812dc42c891b4c2fd

Reply:
Testing is added here https://github.com/elastic/elasticsearch/pull/39756/commits/ce3e69a715af254a64900435c40912b0e58e3c92